zipkin-server/src/test/java/zipkin2/server/internal/eureka/ITZipkinEureka.java

```diff
@@ -0,0 +1,138 @@
+ */
+@SpringBootTest(
+  classes = ZipkinServer.class,
+  webEnvironment = SpringBootTest.WebEnvironment.NONE, // RANDOM_PORT requires spring-web
+  properties = {
+    "server.port=0",
+    "spring.config.name=zipkin-server",
+    "spring.main.banner-mode=off",
+    "zipkin.storage.type=", // cheat and test empty storage type
+    "zipkin.collector.http.enabled=false",
+    "zipkin.query.enabled=false",
+    "zipkin.ui.enabled=false"
+  })
+@Testcontainers
+@TestMethodOrder(OrderAnnotation.class) // so that deregistration tests don't flake the others.
+class ITZipkinEureka {
+
+  @Container static EurekaContainer eureka = new EurekaContainer();
+
+  @Autowired Server zipkin;
+  OkHttpClient client = new OkHttpClient.Builder().followRedirects(false).build();
+
+  /** Get the serviceUrl of the Eureka container prior to booting Zipkin. */
+  @DynamicPropertySource static void propertyOverride(DynamicPropertyRegistry registry) {
+    registry.add("zipkin.discovery.eureka.serviceUrl", eureka::serviceUrl);
+  }
+
+  @Test @Order(1) void registersInEureka() throws Exception {
+    // Notice the app name always coerces to uppercase
+    String json = getEurekaAsString("/apps/ZIPKIN");
+
+    // Make sure the health status is OK
+    assertThat(readString(json, "$.application.instance[0].status"))
+      .isEqualTo("UP");
+
+    String zipkinHostname = zipkin.defaultHostname();
+    int zipkinPort = zipkin.activePort().localAddress().getPort();
+
+    // Note: Netflix/Eureka says use hostname, which can conflict on laptops.
+    // Armeria adopts the spring-cloud-netflix convention shown here.
+    assertThat(readString(json, "$.application.instance[0].instanceId"))
+      .isEqualTo(zipkinHostname + ":zipkin:" + zipkinPort);
+
+    // Make sure the vip address is relevant
+    assertThat(readString(json, "$.application.instance[0].vipAddress"))
+      .isEqualTo(zipkinHostname + ":" + zipkinPort);
+  }
+
+  @Test @Order(2) void deregistersOnClose() {
+    zipkin.close();
+    await().untilAsserted( // wait for deregistration
+      () -> assertThat(getEurekaAsString("/apps"))
+        .isEqualTo(
+          "{\"applications\":{\"versions__delta\":\"1\",\"apps__hashcode\":\"\",\"application\":[]}}"));
+  }
+
+  private String getEurekaAsString(String path) throws IOException {
+    try (Response response = getEureka(path); ResponseBody body = response.body()) {
+      assertThat(response.isSuccessful()).withFailMessage(response.toString()).isTrue();
+      return body != null ? body.string() : "";
+    }
+  }
+
+  private Response getEureka(String path) throws IOException {
+    return client.newCall(new Request.Builder()
+      .url(eureka.serviceUrl() + path)
+      .header("Accept", "application/json") // XML is default
+      .build()).execute();
+  }
+
+  static final class EurekaContainer extends GenericContainer<EurekaContainer> {
```

Comment:
@anuraaga suggested to do this container approach, and yeah it works!

Reply:
Another JFYI :) Now that using Java 11, there would be no cost at all for accessing private members of nested classes (that was improved in Java 10).

Not something to refactor the whole codebase for, but figured it's good to know about.

Reply:
oh cool I didn't know that. will change to help future copy/paste